File: Java/src/main/java/com/azurecosmosdb/cosmoskeyprovider/CosmosRestExecutorOptions.java
package com.azurecosmosdb.cosmoskeyprovider;

import java.time.Instant;
import java.util.Objects;
import java.util.function.IntFunction;

/**
 * Options for Cosmos REST execution/signing.
 * <p>
 * Intentionally small: it mainly exists to allow deterministic signing in unit tests
 * (by controlling the timestamp used for the signature).
 */
public final class CosmosRestExecutorOptions {
  private final String apiVersion;
  private final IntFunction<Instant> utcNowProvider;

  public CosmosRestExecutorOptions() {
    this(CosmosRestRequestSigner.DEFAULT_API_VERSION, attempt -> Instant.now());
  }

  public CosmosRestExecutorOptions(String apiVersion, IntFunction<Instant> utcNowProvider) {
    this.apiVersion = Objects.requireNonNull(apiVersion, "apiVersion");
    this.utcNowProvider = Objects.requireNonNull(utcNowProvider, "utcNowProvider");
  }

  public String apiVersion() {
    return apiVersion;
  }

  public Instant utcNow(int attempt) {
    return utcNowProvider.apply(attempt);
  }
}

File: Csharp/src/CosmosKeyProvider/CosmosRestExecutorOptions.cs
namespace CosmosKeyProvider;

/// <summary>
/// Options for Cosmos REST execution/signing.
/// </summary>
/// <remarks>
/// This is intentionally small: it mainly exists to allow deterministic signing in unit tests
/// (by controlling the timestamp used for the signature).
/// </remarks>
public sealed class CosmosRestExecutorOptions
{
    /// <summary>
    /// Value for the required x-ms-version header.
    /// </summary>
    public string ApiVersion { get; init; } = CosmosRestRequestSigner.DefaultApiVersion;

    /// <summary>
    /// Optional time provider for signing requests. The input is the attempt number (0 = first attempt, 1 = retry).
    /// </summary>
    public Func<int, DateTimeOffset>? UtcNowProvider { get; init; }
}
